File: arduinobot_ws/src/arduinobot_controller/script/angles_converter.py
#! /usr/bin/env python3
import rospy
import math
from arduinobot_controller.srv import AnglesConverter, AnglesConverterResponse

def convert_radians_to_degrees(req):
    # Function that is called every time the service radians_to_degrees is called
    # It receives the Request message as input with the angles in radians
    # and returns the Result message as output with the angles in degrees
    res = AnglesConverterResponse()
    res.base = int(((req.base+(math.pi/2))*180)/math.pi)
    res.shoulder = 180-int(((req.shoulder+(math.pi/2))*180)/math.pi)
    res.elbow = int(((req.elbow+(math.pi/2))*180)/math.pi)
    res.gripper = int(((-req.gripper)*180)/(math.pi/2))
    return res

def convert_degrees_to_radians(req):
    # Function that is called every time the service radians_to_degrees is called
    # It receives the Request message as input with the angles in degrees
    # and returns the Result message as output with the angles in radians
    res = AnglesConverterResponse()
    res.base = ((math.pi*req.base) - ((math.pi/2)*180))/180
    res.shoulder = (((180-req.shoulder)*math.pi)-((math.pi/2)*180))/180
    res.elbow = ((math.pi*req.elbow) - ((math.pi/2)*180))/180
    res.gripper = -((math.pi/2)*req.gripper)/180
    return res


if __name__ == "__main__":
    # Inizialize a ROS node called angles_converter
    rospy.init_node('angles_converter')

    # Inizialize two services for the angle conversions 
    radians_to_degrees = rospy.Service('radians_to_degrees', AnglesConverter, convert_radians_to_degrees)
    degrees_to_radians = rospy.Service('degrees_to_radians', AnglesConverter, convert_degrees_to_radians)

    # keeps the node up and running
    rospy.spin()

File: arduinobot_ws/src/arduinobot_controller/src/angles_converter.cpp
#include "ros/ros.h"
#include "arduinobot_controller/AnglesConverter.h"
#include "math.h"


bool convert_radians_to_degrees(arduinobot_controller::AnglesConverter::Request  &req,
         arduinobot_controller::AnglesConverter::Response &res)
{
    // Function that is called every time the service radians_to_degrees is called
    // It receives the Request message as input with the angles in radians and the Result message Type as input 
    // it fills out the response message with the converted values 
    // and returns the Result message as output with the angles in degrees
    res.base = static_cast<int>(((req.base+(M_PI/2))*180)/M_PI);
    res.shoulder = 180-static_cast<int>(((req.shoulder+(M_PI/2))*180)/M_PI);
    res.elbow = static_cast<int>(((req.elbow+(M_PI/2))*180)/M_PI);
    res.gripper = static_cast<int>(((-req.gripper)*180)/(M_PI/2));
    return true;
}

bool convert_degrees_to_radians(arduinobot_controller::AnglesConverter::Request  &req,
         arduinobot_controller::AnglesConverter::Response &res)
{
    // Function that is called every time the service radians_to_degrees is called
    // It receives the Request message with the angles in degrees and the Result message Type as input 
    // it fills out the response message with the converted values 
    // and returns the Result message as output with the angles in radians
    res.base = ((M_PI*req.base) - ((M_PI/2)*180))/180;
    res.shoulder = (((180-req.shoulder)*M_PI)-((M_PI/2)*180))/180;
    res.elbow = ((M_PI*req.elbow) - ((M_PI/2)*180))/180;
    res.gripper = -((M_PI/2)*req.gripper)/180;
    return true;
}

int main(int argc, char **argv)
{
    // Inizialize a ROS node called angles_converter
    ros::init(argc, argv, "angles_converter");
    ros::NodeHandle n;

    // Inizialize two services for the angle conversions 
    ros::ServiceServer radians_to_degrees = n.advertiseService("radians_to_degrees", convert_radians_to_degrees);
    ros::ServiceServer degrees_to_radians = n.advertiseService("degrees_to_radians", convert_degrees_to_radians);
    
    ROS_INFO("Angles Converter Service Started");

    // keeps the node up and running
    ros::spin();
    return 0;
}